Define the commerce payment transaction entity type.

jkuma opened the pull request:
Here is a first draft of the commerce_payment_transaction entity type shipped with the commerce_payment module. No forms are defined so far, only database model has been implemented.

Please feel free to blame me :)

jkuma committed:
Define the commerce payment transaction entity type.

bojanz commented on modules/payment/src/CommercePaymentTransactionInterface.php:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Provides an interface defining a Commerce Transaction entity.
+ */
+interface CommercePaymentTransactionInterface extends EntityInterface {
```
Shouldn't this extend ContentEntityInterface?

rszrama replied:
Yep. You can reference the Store interfaces / classes if need be.

jkuma replied:
done.

bojanz commented on modules/payment/src/CommercePaymentTransactionInterface.php:
```diff
@@ -0,0 +1,92 @@
+   * @return int
+   *   The entity identifier.
+   */
+  public function id();
```
id(), revision() and friends exist on the parent interface(s). We only need to include what's specific to this entity.

jkuma replied:
Ok, got it.

bojanz commented on modules/payment/src/Entity/CommercePaymentTransaction.php:
```diff
@@ -0,0 +1,270 @@
+ *
+ * @ContentEntityType(
+ *   id = "commerce_payment_transaction",
+ *   label = @Translation("Transaction"),
```
"Payment Transaction" instead of "Transaction" as the label?

rszrama replied:
Agreed.

jkuma replied:
done.

bojanz commented on modules/payment/src/Entity/CommercePaymentTransaction.php:
```diff
@@ -0,0 +1,270 @@
+ *       "edit" = "Drupal\commerce\CommercePaymentTransactionForm",
+ *       "delete" = "Drupal\commerce\Form\CommercePaymentTransactionDeleteForm"
+ *     },
+ *     "translation" = "Drupal\content_translation\ContentTranslationController"
```
The class is now called ContentTranslationHandler

jkuma replied:
done.

bojanz commented on modules/payment/src/Entity/CommercePaymentTransaction.php:
```diff
@@ -0,0 +1,270 @@
+ *     "translation" = "Drupal\content_translation\ContentTranslationController"
+ *   },
+ *   base_table = "commerce_payment_transaction",
+ *   revision_table = "commerce_payment_transaction_revision",
```
This entity type is not revisionable.

bojanz commented on modules/payment/src/Entity/CommercePaymentTransaction.php:
```diff
@@ -0,0 +1,270 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function id() {
```
Already exists on ContentEntityBase, no need to redeclare it. 
(Other core content entities don't seem to do so)

jkuma replied:
yeah, Like you mentioned in your replies above, I'll only declare the methods that are not already implemented in parent classes. Sorry for the mistake.

bojanz commented on modules/payment/src/Entity/CommercePaymentTransaction.php:
```diff
@@ -0,0 +1,270 @@
+  /**
+   * {@inheritdoc}
+   */
+  public function revisionId() {
```
This entity type is not revisionable, and ContentEntityBase has a getRevisionId() that's supposed to be used for fetching the revision id.

bojanz commented:
The entity in this PR has a label, but the D7 entity doesn't have a label. I'm guessing we don't want it?

jkuma committed:
Define the commerce payment transaction entity type. Remove revision table. Complete the entity type interface.

jkuma commented:
Dear DC maintainers,

So I've dropped the revision table and I fixed up the errors you have pointed me. The commerce_payment_transaction interface is now complete. And I agree with you guys, the label property doesn't make sense..... So I dropped it

jkuma committed:
Define the CommercePaymentInformation entity type in order to store payment informations such as credit cards or payment tokens.

rszrama commented on modules/payment/src/CommercePaymentTransactionInterface.php:
```diff
@@ -0,0 +1,228 @@
+
+  /**
+   * Returns the instance identifier for a transaction.
+   *
```
I don't believe we need an instance ID any more, or at least we won't call it the same. This property in Commerce 1.x was just a hash to locate the rule that enabled a payment method so the payment method configuration could be extracted from its action. In Commerce 2.x, we'll be storing payment method configurations via some entity type (it isn't clear to me if it will be a content entity type or configuration entity type). We can change this from instance_id to payment_method_id at the very least; but we can always make this change after we've defined the payment method entity type. : )

jkuma replied:
I don't think it's a blocker point so far, like you said, we can re-adjust the base fields of this entity type afterwards.

rszrama commented on modules/payment/src/CommercePaymentTransactionInterface.php:
```diff
@@ -0,0 +1,228 @@
+   * @return \Drupal\commerce_payment\CommercePaymentTransactionInterface
+   *   The class instance that this method is called on.
+   */
+  public function setAmount($amount);
```
I'm pretty sure we should just make payment transactions fieldable and use a price field this time around. I'm not sure it was a very good idea to include the amount / currency_code as payment transaction properties in 1.x.

jkuma replied:
Yeah, it definitely makes sense to add price field instead of static fields such as "currency_code" and "amount". So I'm gonna drop theses properties and I'll add a TODO for the price field creation.

jkuma commented:
So, Let me summarize things which are included in the commerce_payment module so far:
• The definition of CommercePaymentTransaction entity type. (storing data for payment transactions)
• The definition of CommercePaymentInformation entity type. (storing data around credit cards and payment token, but mostly for credit cards ;) )

bojanz commented:
Merged the module without the payment information interface and entity class, since we agreed to remake them dramatically.
